### PR TITLE
UI Bugfix; Insights navigation causing the browser tab to hang

### PR DESCRIPTION
**Pull request by MartinaeyNL**
Fix for browser crashes we were seeing on the Insights page, when navigating between dashboards.
It was very tricky to discover the cause of this, as it only occurred in production mode,
and the Insights overall is a very complex and intuitive piece of JS code.

I decreased the amount/occurrence of `widgetConfig` corrections, as that seems to be the cause.
Along with that, I've improved logging for the dashboard 'hasChanged' diff checks.

Changelog;
- While in view mode, the 'widget config correction' is now only applied within the preview,
and not exposed to a higher layer. As there is no need to 'save the changes that have been made' for example.
- Updated 'widget config correction' function, to only return a different JS object when changes have been made.
Originally it was returning the same object, but the fields were in a different order; causing unnecessary UI updates.
- Improved logging for `hasChanged` functionality, with the difference between dashboard objects being shared in the console.
- Added new util function using lodash; `difference(obj1, obj2)`
- Added `catch` clause for getting the initial tunnel connection in the `tunnel-widget`.

### CODE REVIEW

**Comment by MartinaeyNL** [comment]
Deployed it on a test server to confirm the issue is fixed 👍

**Comment by MartinaeyNL** [comment]
Confirmed that Insights is working as expected, and nothing breaking is introduced.
Feel free to review.

File: ui/component/or-dashboard-builder/src/index.ts [hunk 1]
```diff
@@ -329,9 +329,30 @@ export class OrDashboardBuilder extends LitElement {
...
-            const templateEqual = Util.objectsEqual(this.currentTemplate, this.initialTemplateJSON ? JSON.parse(this.initialTemplateJSON) : undefined);
-            this.hasChanged = (!dashboardEqual || !templateEqual);
+
+            // Check for dashboard changes
+            const initialDashboard = this.initialDashboardJSON ? JSON.parse(this.initialDashboardJSON) : undefined;
+            const dashboardEqual = Util.objectsEqual(this.selectedDashboard, initialDashboard);
+            if(!dashboardEqual) {
+                this.hasChanged = true;
+                const diff = Util.difference(this.selectedDashboard, initialDashboard);
+                console.debug("Dashboard has detected changes! Diff:", diff);
```
Comment (richturner) [inline ui/component/or-dashboard-builder/src/index.ts:339]: You're doing this diff just to log it to the console? Doesn't seem efficient considering the computation needed

File: ui/component/core/src/util.ts [hunk 2]
```diff
@@ -274,6 +275,22 @@ export function objectsEqual(obj1?: any, obj2?: any, deep: boolean = true): bool
...
     return false;
 }
+/**
+ * Deep diff between two object, using lodash
+ * @param  {Object} object Object compared
+ * @param  {Object} base   Object to compare with
+ * @return {Object}        Return a new object who represent the diff
+ */
+export function difference(object?: any, base?: any): any {
+    function changes(object?: any, base?: any) {
```
Comment (richturner) [inline ui/component/core/src/util.ts:285]: Prefer arrow functions when inlining functions

File: ui/component/or-dashboard-builder/src/index.ts [hunk 3]
```diff
@@ -329,9 +329,30 @@ export class OrDashboardBuilder extends LitElement {
...
+            }
+
+            // Check for template changes
+            else {
+                const initialTemplate = this.initialTemplateJSON ? JSON.parse(this.initialTemplateJSON) : undefined;
+                const templateEqual = Util.objectsEqual(this.currentTemplate, initialTemplate);
+                if(!templateEqual) {
+                    this.hasChanged = true;
+                    const diff = Util.difference(this.currentTemplate, initialTemplate);
+                    console.debug("Template has detected changes! Diff:", diff);
```
Comment (richturner) [inline ui/component/or-dashboard-builder/src/index.ts:349]: Same comment as above

File: ui/component/or-dashboard-builder/src/or-dashboard-widgetcontainer.ts [hunk 4]
```diff
@@ -67,7 +73,7 @@ export class OrDashboardWidgetContainer extends LitElement {
             const oldVal = changedProps.get('widget') as DashboardWidget | undefined;
             const idChanged = oldVal?.id !== this.widget?.id;
             const nameChanged = oldVal?.displayName !== this.widget?.displayName;
-            const configChanged = JSON.stringify(oldVal?.widgetConfig) !== JSON.stringify(this.widget?.widgetConfig);
+            const configChanged = JSON.stringify(oldVal?.widgetConfig) !== JSON.stringify(this.getWidgetConfig());
```
Comment (richturner) [inline ui/component/or-dashboard-builder/src/or-dashboard-widgetcontainer.ts:76]: stringify comparison is not reliable for objects as key order is not guaranteed, that's what `Util.objectsEqual()` method is for

**Review by richturner — CHANGES_REQUESTED** [review]

File: ui/component/core/src/util.ts (hunk 2 above)
Comment (MartinaeyNL) [reply ui/component/core/src/util.ts]: Sure! I adjusted it in the newest commit.

File: ui/component/or-dashboard-builder/src/or-dashboard-widgetcontainer.ts (hunk 4 above)
Comment (MartinaeyNL) [reply ui/component/or-dashboard-builder/src/or-dashboard-widgetcontainer.ts]: Agreed. I adjusted it in the newest commit.

**Comment by MartinaeyNL** [comment]
Thanks for the review again! Made the adjustments you requested 😉

**Review by richturner — APPROVED** [review]

**Pull request merged**